### PR TITLE
Php timezone

### PR DESCRIPTION
**Pull request by jvonau**
### Smoke-tested on which OS or OS's:
http://sprunge.us/kXs9wb

### CODE REVIEW

**Comment by jvonau** [comment]
#69 #1147

**Comment by holta** [comment]
I personally haven't seen evidence that `{{ ansible_date_time.tz }}` is unreliable in recent years.  Of course (A) there are many transliterated synonyms for certain Time Zones, and Ansible's string output can (B) oscillate (e.g. between EDT and EST) depending on summer and winter.  These are more feature than bug (but some may not want these features...)

If a fixed TZ string is truly needed (?) other options include:

- `timedatectl show -p Timezone --value` is rather authoritative.  It uses (1) `/etc/localtime` primarily, or if that's missing it falls back to (2) `/etc/timezone` (if it exists).

- `readlink /etc/localtime | awk -F "zoneinfo/" '{print $2}'` provides the TZ string derived from the symlink.

- `cat /etc/timezone` is no longer authoritative, but can be used as a fallback if absolutely nec.

**Comment by jvonau** [comment]
ansible_date_time.tz is not unreliable just wrong for the purpose.
timedatectl requires the system to be booted, and may not be running. 
readlink works if the system is configured/provisioned.
/etc/timezone is authoritative given timedatectl sets the value or was set by the OS installer at install time.

**Comment by holta** [comment]
> I personally haven't seen evidence that `{{ ansible_date_time.tz }}` is unreliable in recent years.

Let me revise that: Ansible's output above muddles Asia/Tel_Aviv and Asia/Kolkata as IST.  While each is correct, these TZ abbreviations makes it impossible to distinguish e.g. Israel Standard Time (UTC +2) from India Standard Time (UTC +5:30).

And while it's true you can use `{{ ansible_date_time.tz_offset }}` to disambiguate between +2 and + 5:30 &mdash; and you can also use `{{ ansible_date_time.tz_dst }}` to disambiguate between "summer" and "winter" time changes &mdash; this is all very unwieldy.

Conclusion: The original Linux system's TZ (e.g. /etc/localtime) is far more informative, providing all TZ info in 1 single very authoritative place (instead of having to fuss over interpreting 3 separate variables!)

> (A) there are many transliterated synonyms for certain Time Zones, and Ansible's string output can (B) oscillate (e.g. between EDT and EST) depending on summer and winter. These are more feature than bug (but some may not want these features...)

Just FYI it seems likely Ansible is using the same TZ "abbreviator" mechanism as Linux command:

```
date +%Z
```

In short, the above command suffers the exact same problem as Ansible's TZ abbreviation function (both are 1-way functions, that are not reversible, i.e. they produce inconclusive and ambiguous output).

**Comment by holta** [comment]
> /etc/timezone is authoritative given timedatectl sets the value or was set by the OS installer at install time.

In some circumstances.  Those assumptions are not universal.  Some mainline OS installers do not set up /etc/timezone in recent years, ever since /etc/localtime became more authoritative.

**Comment by jvonau** [comment]
> > /etc/timezone is authoritative given timedatectl sets the value or was set by the OS installer at install time.
> 
> In some circumstances. Those assumptions are not universal. Some mainline OS installers do not set up /etc/timezone in recent years, ever since /etc/localtime became more authoritative.

Point one of these OSs out please. I might be open to have local_facts read /etc/timezone first and fallback to using readlink if not found.

**Comment by jvonau** [comment]
timedatectl updates /etc/timezone if the file exists but does not create it if missing, so I will add the fall back using readlink

**Comment by holta** [comment]
I don't recall which Ubuntu/Debian derivative.  But this did happen to me several times in recent years (no /etc/timezone existed).

More important for Matomo: `timedatectl show -p Timezone --value` does the job authoritatively, interpreting the time zone without the need for guesswork or maintenance, regardless whether OS/industry/global standards change.  e.g. it even does the right thing when primary /etc/localtime and secondary /etc/timezone are both missing (in fact this means UTC, according to the latest standard).  With all these low-level grungy corner cases successfully abstracted away by `timedatectl show -p Timezone --value`, I personally would avoid using anything else.

**Comment by jvonau** [comment]
required reading: http://0pointer.de/public/systemd-man/timezone.html https://wiki.archlinux.org/title/System_time
timedatectl does not work without being booted, and would fail in a chrooted environment like pristine image building.

**Comment by holta** [comment]
> timedatectl does not work without being booted, and would fail in a chrooted environment like pristine image building.

Be that as it may, that might not be relevant to @cwivagg (his matomo/tasks/install.yml already relies on systemd / systemctl).

e.g. dynamically invoking `timedatectl show -p Timezone --value` right there from matomo/tasks/install.yml gets the very latest System TZ into PHP for safety (even survives implementers poking around with Time Zones in another window 1min beforehand, on their brand new graphical OS!)

Certainly a substantially friendlier Matomo UX with more meaningful TZ would indeed be great, whether or not squeezed into IIAB 8.0's imminent release :clock11:

**Comment by jvonau** [comment]
You are so full of yourself.
